contract-sdk/types/src/crypto.rs

```diff
@@ -0,0 +1,73 @@
+use std::convert::TryFrom;
+
+use k256::{
+    ecdsa::{recoverable, Signature},
+    EncodedPoint,
+};
+use thiserror::Error;
+
+/// ECDSA recover error.
+#[derive(Error, Debug, cbor::Encode, cbor::Decode)]
+pub enum Error {
+    #[error("malformed input")]
+    MalformedInput,
+    #[error("malformed signature")]
+    MalformedSignature,
+    #[error("signature verification failed")]
+    VerificationFailed,
+}
+
+/// ECDSA public key recovery method.
+pub fn ecdsa_recover(input: &[u8]) -> Result<Vec<u8>, Error> {
+    // [32] hash, [32] r, [32] s, [1] v.
+    if input.len() != 97 {
+        return Err(Error::MalformedInput);
+    }
+
+    let mut msg = [0u8; 32];
+    let mut sig = [0u8; 64];
+    let v = input[96];
+
+    msg[0..32].copy_from_slice(&input[0..32]);
+    sig[0..].copy_from_slice(&input[32..96]);
+
+    let signature = Signature::try_from(&sig[..]).map_err(|_| Error::MalformedSignature)?;
```

Comment:
are various other things like range checks on r and s up to the caller?

Reply:
Does solidity's  `ecrecover` do it? If so, then yes we should probably do it here as well. I think this would ideally be in sync with that, since in most cases this method will be used due to interoperability with ethereum.

Also what are "other things"?


---

Also should probably keep in sync with the EVM's module precomiple: https://github.com/oasisprotocol/oasis-sdk/blob/a7abff8d43ac96c06345a939bf38b15816af7dde/runtime-sdk/modules/evm/src/precompile.rs#L40-L95

Reply:
ok from my reading of the yellowpaper, at least the raw instruction does not. it says the ECRECOVER instruction calls ECDSARECOVER, and the check for canonical `s` is stated outside of that in appendix F.

'other things' I think are just these bounds checks

![image](https://user-images.githubusercontent.com/38229038/135907095-cb647a31-238e-443a-944d-72cbfea2b0bf.png)

the `< secp256k1n` are handled by Signature::try_from I think

Reply:
Thanks, added the `s.is_high()` check i think it makes sense having it.  Should we update the evm precompile as well?

Reply:
Yes I believe it makes sense to have it in the precompile as well, since we do the same check in Ethereum transaction signature verification.

Reply:
Done: https://github.com/oasisprotocol/oasis-sdk/pull/505